### DISCUSSION TITLE
How to Handle Errors for ActionForm

### DISCUSSION THREAD

**Discussion by jeorb** [leptos-rs/leptos#673] [Q&A] [answered]
Is it possible to have a view handle errors from the response to an ActionForm? I would like to create some forms that update with feedback to the user when there is an error from the server.

I tested the session_auth_axum example and it looks like the there is some intention to show login errors in the header but the error message never seems to show:

`leptos/examples/session_auth_axum/src/todo.rs lines 168-172`
~~~
Err(e) => view! {cx,
    <A href="/signup">"Signup"</A>", "
    <A href="/login">"Login"</A>", "
    <span>{format!("Login error: {}", e.to_string())}</span>
}.into_view(cx),
~~~

When I log in using an invalid username or password, I see the 500 error is returned to the browser with a body like `error running server function: Password does not match.` but that response doesn't trigger an update to the page.

**Comment by gbj** [comment, marked as answer]
Yep so there are a handful of different things here:
1. We let you pass an `error` signal into a `<Form/>` but not an `<ActionForm/>`, which will hold any errors that come up during the actual submission process... I should probably add that to `<ActionForm/>` too.
2. The lines you quote are a little misleading as they're actually on the resource reading the current user — the error should probably be something like "Error getting user info." This would come up if, for example, the network request to the server to get the latest user data failed because you went offline.
3. For error handling with an actual form you can check the `.value()` of the action you're giving to the `<ActionForm/>`. `.value()` returns a signal that contains the latest response/value that the form has received. So the `<Login/>` component could do
```rust
    let response = action.value();
    let error = move || match response.get() {
        Some(Err(e)) => Some(e.to_string()),
        _ => None,
    };
```
And then like a `{error.map(|e| view! { cx, <p class="error">{e}</p> })}` somewhere.

(When I try this I get a deserialization error because the form is expecting JSON, not a string with the error message, but that's a separate issue!)

Some of these should probably spawn actual issues or feature requests; feel free to open issues if you run into problems like this. I'll do that now for the deserialization thing.
